pkg/ipfs/registry.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		return nil, "", 0, err
+	}
+	f, err := s.getFile(targetCID)
+	if err != nil {
+		return nil, "", 0, err
+	}
+	ra, ok := f.(interface {
+		io.ReaderAt
+	})
+	if !ok {
+		return nil, "", 0, fmt.Errorf("ReaderAt is not implemented")
+	}
+	size, err = f.Size()
+	if err != nil {
+		return nil, "", 0, err
+	}
+	return &readSeekCloser{io.NewSectionReader(ra, 0, size), f.Close}, detectMediaType(targetDesc), targetDesc.Size, nil
+}
+
+func (s *server) getFile(c cid.Cid) (files.File, error) {
+	n, err := s.api.Unixfs().Get(context.Background(), ipath.IpfsPath(c)) // only IPFS CID is supported
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file %q: %v", c.String(), err)
+	}
+	return files.ToFile(n), nil
+}
+
+func (s *server) resolveCIDOfRootBlob(c cid.Cid) (cid.Cid, ocispec.Descriptor, error) {
+	rc, err := s.getFile(c)
+	if err != nil {
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	defer rc.Close()
+	var desc ocispec.Descriptor
+	if err := json.NewDecoder(rc).Decode(&desc); err != nil {
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	c, err = getIPFSCID(desc)
+	if err != nil {
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	return c, desc, nil
+}
+
+func (s *server) resolveCIDOfDigest(dgst digest.Digest, desc ocispec.Descriptor) (cid.Cid, ocispec.Descriptor, error) {
+	c, err := getIPFSCID(desc)
+	if err != nil {
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	if desc.Digest == dgst {
+		return c, desc, nil // hit
+	}
+	if !images.IsManifestType(desc.MediaType) && !images.IsIndexType(desc.MediaType) {
+		return cid.Cid{}, ocispec.Descriptor{}, fmt.Errorf("not found") // no child blob is contained
+	}
+	b, err := s.getFile(c)
+	if err != nil {
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	descs, err := getChildrenOfManifestOrIndex(b)
+	if err != nil {
+		b.Close()
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	if err := b.Close(); err != nil {
+		return cid.Cid{}, ocispec.Descriptor{}, err
+	}
+	var allErr error
+	for _, desc := range descs {
+		gotCID, gotDesc, err := s.resolveCIDOfDigest(dgst, desc)
+		if err != nil {
+			allErr = multierror.Append(allErr, err)
+			continue
+		}
+		return gotCID, gotDesc, nil
+	}
+	if allErr == nil {
+		return cid.Cid{}, ocispec.Descriptor{}, fmt.Errorf("not found")
+	}
+	return cid.Cid{}, ocispec.Descriptor{}, allErr
+}
+
+// OCIBlob is an OCI object which can contain children descriptors.
+// OCI manifest or index can be persed to this struct.
+type OCIBlob struct {
+	// Config references a configuration object for a container, by digest.
+	// The referenced configuration object is a JSON blob that the runtime uses to set up the container.
+	Config ocispec.Descriptor `json:"config"`
+
+	// Layers is an indexed list of layers referenced by the manifest.
+	Layers []ocispec.Descriptor `json:"layers"`
+
+	// Manifests references platform specific manifests.
+	Manifests []ocispec.Descriptor `json:"manifests"`
+}
```

Comment:
We can make it mediatype-aware after we bump up github.com./opencontainers/image-spec to v1.0.2.

Reply:
Fixed this by leveraging `images.Children()` of containerd lib.